### PR TITLE
Set a specific SKU in the product context

### PR DESCRIPTION
**Pull request by Git-the-Sanz**
#### What problem is this solving?

There was no way of specifying which SKU should be the initial selection in product summaries (Both sliders and galleries).
We are now letting the client specify a logic to then use it for the product-context, which in turn will provide the selected item property to it's childrens.

#### How to test it?

[CHEAPEST](https://cheapest--motoroladevelopment.myvtex.com/)
(You will see the cheapest products selected)

[SPECIFICATION](https://specification--motoroladevelopment.myvtex.com/Smartphones/Moto-G-Family?page=1)
(Look for MOTO G7 PLUS, you will see that the SKU No. 35 is selected, thanks to it's product property called `DefaultSKU`)

#### Screenshots or example usage:
For i.e:
By passing "SPECIFICATION" the SKU No. 35 was automatically selected in the Gallery

item

#### Describe alternatives you've considered, if any.

We know that the Admin SKU Position exists, but as of today, this is not working (Confirmed by Catalog team)

#### Related to / Depends on

[Search result PR](https://github.com/vtex-apps/search-result/pull/536)

### CODE REVIEW

File: react/utils/normalize.ts
Comment (icazevedo) [inline react/utils/normalize.ts:136]: I think you could simplify this by doing something like:

```suggestion
    const price = getPriceByCondition(curr, condition)
    const currentBestPrice = getPriceByCondition(acc, condition)
   
    if (condition === 'expensive') {
      return price > currentBestPrice ? curr : acc
    }

    return price < currentBestPrice ? curr : acc
  })
```

File: react/utils/normalize.ts
Comment (icazevedo) [inline react/utils/normalize.ts:152]: I'm really not sure about this. Imagine the scenario: the client wants some products to follow their default sku specification, and the rest to be ordered following `PRICE_ASC`. It wouldn't be possible with this code, since if products don't have a default SKU specified, the selected SKU will always be the first one, and there'll be no way of changing that.

What I propose is: `getSpecificSKU` or `findSKUBySpecification` (however you name it) should be the default choosing option, and should always apply if such a specification is set for that product. It shouldn't be a choice on the theme. If the client added the specification to the product, that's what they want, right?

That wouldn't change the behavior of current stores, since they don't have any product with that specification set. This way, after not finding any SKU with that specification, the code would use the specified `PreferredSKU` option, which by default would be `FIRST_AVAILABLE`, thus not changing how the store currently works. What do you think?

**Review by icazevedo — CHANGES_REQUESTED** [review]

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:15]: Added

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:103]: Added

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:210]: Added

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:216]: Added

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:219]: Added

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:136]: Added

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:124]: Added

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:89]: Added

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:97]: I added the null check; but also refactored the methods to get seller/s stock. It should now look cleaner.

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:181]: Added

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:152]: A really good idea @icazevedo!
I made it so it first checks if the property exists, and if so attempts to find the desire SKU ID.
If that fails, it will default to the condition supplied.

**Comment by Git-the-Sanz** [comment]
@icazevedo Thank you for your input.
- I typed everything; except the property `normalizedProduct.sku` (I could't manage to do it, the object is particularly weird; and it's used and mutated-around inside product-summary)
- I renamed some functions. They now reflect better what they do.
- I refactored the reduce function. Since the beginning I felt that this approach was barely performant. I now only run the price lookup function once for each SKU, and then just return the best one using the index from max/min.
- Regarding why I'm using .some() I only care if at least 1 seller has stock in a particular Item, if not, it will help us avoid overhead. That's the difference with the other applications.
- Regarding your comment about the Specification. I implemented that one. We first check if the property exists, if it does, we will attempt to retrieve it; if the property is incorrectly valued, we instead attempt to use the switch to find a suitable one.

Thank you for all your inputs!

File: react/utils/normalize.ts
Comment (igorbrasileiro) [inline react/utils/normalize.ts:243]: Just a doubt, why do you need to type as any?

File: react/utils/normalize.ts
Comment (igorbrasileiro) [inline react/utils/normalize.ts:236]:
```suggestion
 */
```

File: react/utils/normalize.ts
Comment (igorbrasileiro) [inline react/utils/normalize.ts:214]: What do you think about these suggestions?
```suggestion
        items[items.length - 1]
```
or 
```ts
case 'LAST_AVAILABLE':
    const reservedItems = [...items].reverse()
    return (reversedItems.find(getAvailableProduct) ?? reversedItems[0])
```

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:243]: I'm sending this as any because by typing the product in the function args, and referencing it again in the normalizeProduct, it would trigger the type check.

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts]: Added

File: react/utils/normalize.ts
Comment (igorbrasileiro) [reply react/utils/normalize.ts:243]: Can you share where it will trigger the type check? It's breaking the typescript build? I can try to help you.a

File: react/utils/normalize.ts
Comment (igorbrasileiro) [inline react/utils/normalize.ts:176]: Ain't an English expert, but Grammarly warned about it.
```suggestion
 * Attempts to retrieve the specific SKU using a custom Product (field)
 * in the Catalog. The specification has the value of the SKU ID
 * that will be looking for. If not found for any reason, we default
 * to finding the preferred SKU
 * @param items an array of SKUs
 * @param defaultSKUspec the ID used to find the SKU
 * @param preferenceFallback the logic for finding an SKU if the ID fails
```

File: react/utils/normalize.ts
Comment (igorbrasileiro) [inline react/utils/normalize.ts:235]: Grammarly attacks again haha! Feel free to argue against 
```suggestion
 * @description
 * This method is responsible for constructing the product type for the product-context.
 * The property `product.sku` is built by these specifications
```

File: CHANGELOG.md
Comment (igorbrasileiro) [inline CHANGELOG.md:10]:
```suggestion
- The ability to dynamically choose the SKU that will be initially selected

```

**Review by igorbrasileiro — APPROVED** [review]
Hey @Git-the-Sanz, marvelous documentation! Great job 👏 🚀 .
I left some comments, can you discuss them?
Also, remember to fix the changelog

**Review by Git-the-Sanz — COMMENTED** [review]
Added all requested changes

**Review by Git-the-Sanz — COMMENTED** [review]
Ready for review

File: .eslintrc
Comment (victorhmp) [inline .eslintrc:6]: Why was this change necessary?

File: docs/ProductSummaryList.md
Comment (victorhmp) [inline docs/ProductSummaryList.md:68]: Is there a recipe for this already? If so, update this link to point directly to the recipe, instead of to the recipe list.

Also, I think we're supposed to be linking to developers.vtex.com instead of vtex.io/docs 😄

File: docs/ProductSummaryList.md
Comment (victorhmp) [inline docs/ProductSummaryList.md:67]: I don't quite understand what you mean here. The default component behavior has changed?

File: react/utils/normalize.ts
Comment (victorhmp) [inline react/utils/normalize.ts:100]: Small nitpick, but we prefer always writing `if`s in this form to make them easier to read.

```suggestion
  if (sellers.length === 1) {
    return sellers[0].commertialOffer.Price
  }
```

File: react/utils/normalize.ts
Comment (victorhmp) [inline react/utils/normalize.ts:1]: This file has become way more complex than before, no problem with that. But could you write some tests for these new functions?

Since they are simple TS functions, you wouldn't need to use anything besides Jest to do so :)

**Review by victorhmp — CHANGES_REQUESTED** [review]

File: docs/ProductSummaryList.md
Comment (Git-the-Sanz) [reply docs/ProductSummaryList.md:68]: The education team will create the recipe once this is approved.
Regarding the url, I don't see the Recipes subdirectory in developers.vtex.com

File: docs/ProductSummaryList.md
Comment (Git-the-Sanz) [reply docs/ProductSummaryList.md:67]: Yes, previously the initially selected SKU was the first available, now it's the product specification

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:1]: I'm not familiar with it, but I'll see what I can do

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts]: Added

**Review by Git-the-Sanz — COMMENTED** [review]
Got tests for the 'PRICE_ASC' and for when the 'SPECIFICATION' is in place. Also fixed other test using the mapCatalogProductToProductSummary method

File: .eslintrc
Comment (victorhmp) [reply .eslintrc:6]: Yep, that's fine then :)

File: docs/ProductSummaryList.md
Comment (victorhmp) [reply docs/ProductSummaryList.md:67]: But wouldn't this affect all stores then? If so this would be a breaking change

File: docs/ProductSummaryList.md
Comment (victorhmp) [reply docs/ProductSummaryList.md:68]: Then I think you could remove this part for now

**Review by Git-the-Sanz — COMMENTED** [review]
Updated readme

**Review by victorhmp — APPROVED** [review]
Awesome! 👏🏻

File: react/utils/normalize.ts
Comment (Git-the-Sanz) [reply react/utils/normalize.ts:245]: If you remove this type assignment, you'll see that on line 278 the property `sku` does not exist on `normalizedProduct` because when doing this: `const normalizedProduct = { ...product }` TS assigns the original type `ProductTypes.Product` to the new Object. 

You could intersect product and SKU, but you'll also need to type a new referenceId and the new images properties, this is way too much effort for so little gain...

**Review by Git-the-Sanz — COMMENTED** [review]
Awaiting approval

File: react/utils/normalize.ts
Comment (igorbrasileiro) [reply react/utils/normalize.ts:245]: You're right, but have you ever tried to do something like that?

```suggestion
  const normalizedProduct: Product = { ...product }
```
Product comes from [product-summary-context](https://github.com/vtex-apps/product-summary-context/blob/master/react/ProductSummaryTypes.ts#L14)

If it doesn't work, you can try something like that:

```ts
const normalizedProduct: ProductTypes.Product & { sku?: SingleSKU} = { ...product }
```
[SingleSKU](https://github.com/vtex-apps/product-summary-context/blob/master/react/ProductSummaryTypes.ts#L52)

Comment (igorbrasileiro) [reply react/utils/normalize.ts:245]: We need to have everything typed, help others developers, and prevents us from miswriting code.

Comment (Git-the-Sanz) [reply react/utils/normalize.ts:245]: ex

No because now this doesn't work... really where's the gain of typing something like this?

Comment (Git-the-Sanz) [reply react/utils/normalize.ts:245]: The type is only used here, and nowhere else, and If we needed to arrange it to look like the original types, we would need to change the previous functions not assigned to this PR.

Comment (Git-the-Sanz) [reply react/utils/normalize.ts:245]: This wasn't even typed before! I'm just getting the flak the because someone else didn't do it.

**Review by icazevedo — APPROVED** [review]

**Review by igorbrasileiro — APPROVED** [review]

File: react/utils/normalize.ts
Comment (victorhmp) [reply react/utils/normalize.ts:245]: @Git-the-Sanz I'm going to merge this PR now, but just to let you know for the future, this is definitely **not** how engineering at VTEX is treated. With this mindset of "This wasn't even typed before! I'm just getting the flak because someone else didn't do it. This "normalizedProduct" is not of my doing.". If we're making changes to a bit of code that you can improve, by typing, linting, simplifying... whatever it is, we do it. We're all working to make our products better, and we help each other looking over the quality of the code instead of blaming someone else for now having done something before.

**Pull request merged**
